core/src/main/java/org/apache/calcite/sql/validate/DelegatingScope.java

```diff
@@ -214,6 +216,47 @@ public SqlValidator getValidator() {
     return validator;
   }
 
+  /**
+   *
+   * @param map A map from table names to {@code ScopeChild} objects, which
+   *            represents all possible tables which a particular column
+   *            could come from.
+   * @return    A new table name map with unknown tables filtered out
+   *            (unless only unknown tables are present, in which case it
+   *            leaves an arbitrary unknown table which is assumed to contain
+   *            the column we are trying to resolve).
+   *            This ensures that columns are not matched to unknown tables if
+   *            they can instead be matched to tables with known schemas.
+   */
+  public Map<String, ScopeChild> filterTableNames(Map<String, ScopeChild> map) {
+    Map<String, ScopeChild> filteredMap = new HashMap<String, ScopeChild>();
+    for (String key : map.keySet()) {
+      if (!containsUnknownNamespace(map.get(key))) {
+        filteredMap.put(key, map.get(key));
+      }
+    }
+    // If original map consists only of unknown tables,
```

Comment:
nit: This line looks like it's broken before the 80 character cut-off and could have had the word "arbitrarily" without going over. Please adjust this comment (and others throughout the PR) such that a line contains as many words as possible (without going over 80 characters) before going to the next line.